File: app/src/main/java/com/walfud/contactsync_android/main/MainPresenterImpl.java
package com.walfud.contactsync_android.main;

import android.util.Log;

import com.walfud.contactsync_android.ContactSyncApplication;
import com.walfud.contactsync_android.SyncMutation;
import com.walfud.contactsync_android.model.ContactRealm;
import com.walfud.contactsync_android.service.contact.ContactModel;
import com.walfud.contactsync_android.service.contact.ContactService;
import com.walfud.contactsync_android.service.network.NetworkService;
import com.walfud.contactsync_android.service.user.UserService;
import com.walfud.contactsync_android.type.ContactInputType;
import com.walfud.walle.lang.ObjectUtils;

import java.util.List;
import java.util.stream.Collectors;

import io.reactivex.SingleObserver;
import io.reactivex.disposables.Disposable;
import io.realm.Realm;

/**
 * Created by walfud on 2017/4/20.
 */

public class MainPresenterImpl implements MainPresenter {

    public static final String TAG = "MainPresenterImpl";

    private MainView mMainView;
    private UserService mUserService;
    private NetworkService mNetworkService;

    public MainPresenterImpl(MainView mainView, UserService userService, NetworkService networkService) {
        mMainView = mainView;
        mUserService = userService;
        mNetworkService = networkService;
    }

    @Override
    public void onLogin(String oid, String accessToken, String refreshToken) {
        mUserService.changeUser(oid);
        mUserService.setToken(accessToken);
    }

    @Override
    public void onRefresh() {
        List<MainView.ViewContactData> local = ContactService.getContactList(ContactSyncApplication.getAppContext()).stream()
                .map(contactModel -> {
                    Realm realm = Realm.getDefaultInstance();
                    ContactRealm contactRealm = realm.where(ContactRealm.class)
                            .equalTo("localId", contactModel.id)
                            .findFirst();

                    MainView.ViewContactData viewContactData = new MainView.ViewContactData();
                    if (contactRealm == null) {
                        // New contact
                        viewContactData.name = contactModel.displayName;
                        viewContactData.phoneList = contactModel.phoneList.stream()
                                .map(phoneModel -> phoneModel.number)
                                .collect(Collectors.toList());
                        viewContactData.status = MainView.ViewContactData.STATUS_LOCAL_ONLY;
                    } else {
                        // TODO: show difference
                        viewContactData.name = contactModel.displayName;
                        viewContactData.phoneList = contactModel.phoneList.stream()
                                .map(phoneModel -> phoneModel.number)
                                .collect(Collectors.toList());
                        viewContactData.status = MainView.ViewContactData.STATUS_CHANGED;
                    }

                    return viewContactData;
                })
                .collect(Collectors.toList());
        mMainView.show(local);
    }

    @Override
    public void onSync() {
        List<ContactModel> clientContactList = ContactService.getContactList(ContactSyncApplication.getAppContext());
        mNetworkService.sync(mUserService.getToken(), clientContactList.stream()
                .map(contactModel -> {
                    Realm realm = Realm.getDefaultInstance();
                    ContactRealm contactRealm = realm.where(ContactRealm.class)
                            .equalTo("localId", contactModel.id)
                            .findFirst();

                    ContactInputType.Builder contactBuilder = ContactInputType.builder();
                    if (contactRealm != null) {
                        contactBuilder.id(contactRealm.id);
                    }
                    contactBuilder.name(contactModel.displayName)
                            .phones(contactModel.phoneList.stream()
                                    .map(phoneModel -> phoneModel.number)
                                    .collect(Collectors.toList()))
                            .modify_time(contactModel.modifyTime)
                            .is_deleted(contactModel.isDeleted);

                    return contactBuilder.build();
                })
                .collect(Collectors.toList()))
                .subscribe(new SingleObserver<SyncMutation.Data>() {
                    @Override
                    public void onSubscribe(Disposable d) {

                    }

                    @Override
                    public void onSuccess(SyncMutation.Data data) {
                        List<SyncMutation.Data.Contact> serverContactList = data.sync().contacts();
                        if (serverContactList.size() < clientContactList.size()) {
                            throw new RuntimeException("server response error");
                        }

                        //
                        Realm realm = Realm.getDefaultInstance();
                        realm.beginTransaction();
                        realm.deleteAll();
                        for (int i = 0; i < serverContactList.size(); i++) {
                            SyncMutation.Data.Contact serverContact = serverContactList.get(i);
                            if (i < clientContactList.size()) {
                                // Client vs. Server
                                ContactModel clientContact = clientContactList.get(i);
                                if (!ObjectUtils.isEqual(clientContact.displayName, serverContact.name())
                                        || !ObjectUtils.isEqual(clientContact.phoneList, serverContact.phones(), (clientPhone, serverPhone) -> ObjectUtils.isEqual(clientPhone.number, serverPhone) ? 0 : -1)
                                        || clientContact.isDeleted != serverContact.is_deleted()) {
                                    if (!serverContact.is_deleted()) {
                                        ContactService.update(ContactSyncApplication.getAppContext(), clientContact.id, serverContact.name(), serverContact.phones());
                                        Log.d(TAG, String.format("modify: %10s, %s", serverContact.name(), serverContact.phones().toString()));

                                    } else {
                                        ContactService.delete(ContactSyncApplication.getAppContext(), clientContact.id);
                                        Log.d(TAG, String.format("del: %10s, %d", serverContact.name(), clientContact.id));
                                    }
                                }
                                ContactRealm contactRealm = ContactRealm.valueOf(realm, serverContact);
                                contactRealm.localId = clientContact.id;
                            } else {
                                // Only Server
                                long localId = ContactService.insert(ContactSyncApplication.getAppContext(), serverContact.name(), serverContact.phones());
                                ContactRealm contactRealm = ContactRealm.valueOf(realm, serverContact);
                                contactRealm.localId = localId;
                                Log.d(TAG, String.format("add: %10s, %s", serverContact.name(), serverContact.phones().toString()));
                            }
                        }
                        realm.commitTransaction();
                    }

                    @Override
                    public void onError(Throwable e) {
                        mMainView.error(e.getMessage());
                    }
                });
    }
}

File: app/src/main/java/com/walfud/contactsync_android/main/MainPresenterImpl.kt
package com.walfud.contactsync_android.main

import android.widget.Toast
import com.walfud.contactsync_android.ContactQuery
import com.walfud.contactsync_android.ContactSyncApplication
import com.walfud.contactsync_android.R
import com.walfud.contactsync_android.appContext
import com.walfud.contactsync_android.main.MainContract.MainPresenter
import com.walfud.contactsync_android.main.MainContract.MainView
import com.walfud.contactsync_android.main.MainContract.MainView.ViewContactData
import com.walfud.contactsync_android.service.network.NetworkService
import com.walfud.contactsync_android.service.user.UserService
import com.walfud.contactsync_android.type.ContactInputType
import com.walfud.walle.android.contact.ContactUtils
import com.walfud.walle.lang.ObjectUtils
import io.reactivex.Observable
import io.reactivex.Observer
import io.reactivex.Single
import io.reactivex.SingleObserver
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers

/**
 * Created by walfud on 2017/4/20.
 */

class MainPresenterImpl(val mMainView: MainView) : MainPresenter {

    var contacts: List<ViewContactData> = listOf()

    override fun onLogin(oid: String, accessToken: String, refreshToken: String) {
        UserService.changeUser(oid)
        UserService.token = accessToken
    }

    override fun onRefresh() {
        mMainView.show(contacts)
    }

    override fun onSync() {
    }

    override fun onDownload() {
        Observable.just(0)
                .observeOn(Schedulers.io())
                .map {
                    // Local contacts
                    ContactUtils.getContactList(ContactSyncApplication.appContext)
                            .filter { !it.isDeleted }
                            .map { contactModel ->
                                ViewContactData(contactModel.displayName!!, contactModel.phoneList?.map { phoneModel -> phoneModel.number!! } ?: arrayListOf(), ViewContactData.STATUS_LOCAL_ONLY)
                            }
                }
                .observeOn(AndroidSchedulers.mainThread())
                .map { localContacts ->
                    contacts = localContacts
                    onRefresh()
                    contacts
                }
                .concatMap { localContacts ->
                    // Remote
                    Observable.create<ContactQuery.Data.Contact>({ e ->
                        NetworkService.download(UserService.token).contacts()!!.forEach(e::onNext)
                        e.onComplete()
                    })
                            .subscribeOn(Schedulers.io())
                            .filter({ serverContact ->
                                localContacts.find { ObjectUtils.isEqual(it.name + it.phoneList.toString(), serverContact.name() + serverContact.phones().toString()) } == null
                            })
                            .observeOn(AndroidSchedulers.mainThread())
                            .map { contact ->
                                ViewContactData(contact.name()!!, contact.phones()!!, ViewContactData.STATUS_REMOTE_ONLY)
                            }
                }
                .observeOn(Schedulers.io())
                .map { contact ->
                    ContactUtils.insert(appContext, contact.name, contact.phoneList)
                    contact
                }
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(object : Observer<ViewContactData> {
                    override fun onSubscribe(d: Disposable?) {
                        mMainView.loading(true)
                    }

                    override fun onNext(contact: ViewContactData) {
                        contacts = contacts.plus(contact)
                        onRefresh()
                    }

                    override fun onComplete() {
                        mMainView.loading(false)
                        Toast.makeText(appContext, R.string.main_finish, Toast.LENGTH_SHORT).show()
                    }

                    override fun onError(e: Throwable) {
                        mMainView.loading(false)
                        Toast.makeText(appContext, e.message ?: "", Toast.LENGTH_SHORT).show()
                    }
                })
    }

    override fun onUpload() {
        Single.just(0)
                .observeOn(Schedulers.io())
                .map {
                    NetworkService.upload(UserService.token, contacts.map { ContactInputType.builder().id("").name(it.name).phones(it.phoneList).build() }).upload()!!
                }
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(object : SingleObserver<Boolean> {
                    override fun onSubscribe(d: Disposable) {
                        mMainView.loading(true)
                    }

                    override fun onSuccess(t: Boolean) {
                        mMainView.loading(false)
                        Toast.makeText(appContext, "Success", Toast.LENGTH_SHORT).show()
                    }

                    override fun onError(e: Throwable) {
                        mMainView.loading(false)
                        Toast.makeText(appContext, e.message, Toast.LENGTH_SHORT).show()
                    }
                })
    }

}
